sycl/unittests/SYCL2020/KernelID.cpp

```diff
@@ -0,0 +1,277 @@
+  }
+  static constexpr const char *getName() { return "KernelID_TestKernel1"; }
+  static constexpr bool isESIMD() { return false; }
+  static constexpr bool callsThisItem() { return false; }
+  static constexpr bool callsAnyThisFreeFunction() { return false; }
+};
+
+template <> struct KernelInfo<TestKernel2> {
+  static constexpr unsigned getNumParams() { return 0; }
+  static const kernel_param_desc_t &getParamDesc(int) {
+    static kernel_param_desc_t Dummy;
+    return Dummy;
+  }
+  static constexpr const char *getName() { return "KernelID_TestKernel2"; }
+  static constexpr bool isESIMD() { return false; }
+  static constexpr bool callsThisItem() { return false; }
+  static constexpr bool callsAnyThisFreeFunction() { return false; }
+};
+
+template <> struct KernelInfo<TestKernel3> {
+  static constexpr unsigned getNumParams() { return 0; }
+  static const kernel_param_desc_t &getParamDesc(int) {
+    static kernel_param_desc_t Dummy;
+    return Dummy;
+  }
+  static constexpr const char *getName() { return "KernelID_TestKernel3"; }
+  static constexpr bool isESIMD() { return false; }
+  static constexpr bool callsThisItem() { return false; }
+  static constexpr bool callsAnyThisFreeFunction() { return false; }
+};
+
+} // namespace detail
+} // namespace sycl
+} // __SYCL_INLINE_NAMESPACE(cl)
+
+static sycl::unittest::PiImage
+generateDefaultImage(std::initializer_list<std::string> Kernels) {
+  using namespace sycl::unittest;
+
+  PiPropertySet PropSet;
+
+  std::vector<unsigned char> Bin{0, 1, 2, 3, 4, 5}; // Random data
+
+  PiArray<PiOffloadEntry> Entries = makeEmptyKernels(Kernels);
+
+  PiImage Img{PI_DEVICE_BINARY_TYPE_SPIRV,            // Format
+              __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64, // DeviceTargetSpec
+              "",                                     // Compile options
+              "",                                     // Link options
+              std::move(Bin),
+              std::move(Entries),
+              std::move(PropSet)};
+
+  return Img;
+}
+
+static sycl::unittest::PiImage Imgs[2] = {
+    generateDefaultImage({"KernelID_TestKernel1", "KernelID_TestKernel3"}),
+    generateDefaultImage({"KernelID_TestKernel2"})};
+static sycl::unittest::PiImageArray<2> ImgArray{Imgs};
+
+TEST(KernelID, AllProgramKernelIds) {
+  std::vector<sycl::kernel_id> AllKernelIDs = sycl::get_kernel_ids();
+
+  // sycl::get_kernel_ids may pick up kernels from other unit tests, so there
+  // must be *at least* the 3 from this test.
+  EXPECT_GE(AllKernelIDs.size(), 3u);
+
+  sycl::kernel_id TestKernel1ID = sycl::get_kernel_id<TestKernel1>();
+  sycl::kernel_id TestKernel2ID = sycl::get_kernel_id<TestKernel2>();
+  sycl::kernel_id TestKernel3ID = sycl::get_kernel_id<TestKernel3>();
+
+  for (sycl::kernel_id TestKernelID :
+       {TestKernel1ID, TestKernel2ID, TestKernel3ID}) {
+    bool Found = false;
+    for (sycl::kernel_id KernelID : AllKernelIDs) {
+      Found = TestKernelID == KernelID || Found;
```

Comment:
Can we early exit if the kernel id is found? I would expect that the number of kernel ids will be growing each time we add a new test case.
Also using `std::find` should work well here I believe.

Reply:
Absolutely!

> Also using std::find should work well here I believe.

I convinced myself that std::any_of made more sense, but revisiting that thought I think you're right that std::find is better. I'll change it.